tools/cacheload/cacheload.go

```diff
@@ -0,0 +1,214 @@
+package main
+
+import (
+	"bytes"
+	"context"
+	"flag"
+	"fmt"
+	"io"
+	"math"
+	"math/rand"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/buildbuddy-io/buildbuddy/server/remote_cache/cachetools"
+	"github.com/buildbuddy-io/buildbuddy/server/remote_cache/digest"
+	"github.com/buildbuddy-io/buildbuddy/server/util/grpc_client"
+	"github.com/buildbuddy-io/buildbuddy/server/util/log"
+	"golang.org/x/sync/errgroup"
+	"golang.org/x/time/rate"
+	"google.golang.org/grpc/metadata"
+
+	repb "github.com/buildbuddy-io/buildbuddy/proto/remote_execution"
+	bspb "google.golang.org/genproto/googleapis/bytestream"
+)
+
+var (
+	cacheTarget  = flag.String("cache_target", "localhost:1985", "Cache target to connect to.")
+	writeQPS     = flag.Uint("write_qps", 100, "How many queries per second to attempt to write.")
+	readQPS     = flag.Uint("read_qps", 1000, "How many queries per second to attempt to read.")
+	instanceName = flag.String("instance_name", "loadtest", "An optional Remote Instance name.")
+	apiKey       = flag.String("api_key", "", "An optional API key to use when reading / writing data.")
+
+	realisticBlobSizes = flag.Bool("realistic_blob_sizes", true, "If true, use realistic blob sizes, ignoring blob_size flag.")
+	blobSize           = flag.Int64("blob_size", 100000, "Num bytes (max) of blob to send/read.")
+	recycleRate        = flag.Float64("recycle_rate", .10, "If true, re-queue digests for read after reading")
+)
+
+const (
+	byteStreamRead   = "google.bytestream.ByteStream/Read"
+	byteStreamWrite  = "google.bytestream.ByteStream/Write"
+	findMissingBlobs = "build.bazel.remote.execution.v2.ContentAddressableStorage/FindMissingBlobs"
+)
+
+var (
+	digestGenerator   *digest.Generator
+	mu                sync.Mutex
+)
+
+var (
+	// Data computed by sampling stored cache blob sizes.
+	histBuckets     = []int{1, 10, 100, 1000, 10000, 100000, 1000000, 10000000, 100000000}
+	histCounts      = []int{23, 33611, 33498, 20473, 10036, 3265, 504, 62}
+	histCountsTotal int
+)
+
+func init() {
+	for _, c := range histCounts {
+		histCountsTotal += c
+	}
+}
+
+type Counter struct {
+	c uint64
+}
+func (c *Counter) Get() uint64 {
+	return atomic.LoadUint64(&c.c)
+}
+func (c *Counter) Inc() uint64 {
+	return atomic.AddUint64(&c.c, 1)
+}
+
+func randRange(low, high int) int64 {
+	i := int64(rand.Intn(high-low+1) + low)
+	return i
+}
+
+func randomBlobSize() int64 {
+	if !*realisticBlobSizes {
+		return *blobSize
+	}
+	n := rand.Intn(histCountsTotal)
+	var sumTotal, low, high int
+	for i, c := range histCounts {
+		sumTotal += c
+		high = histBuckets[i+1]
+		if n < sumTotal {
+			return randRange(low, high)
+		}
+		low = histBuckets[i+1]
+	}
+	return randRange(histBuckets[len(histBuckets)-2], histBuckets[len(histBuckets)-1])
+}
+
+var generatorMu = sync.Mutex{}
+
+func newRandomDigestBuf(sizeBytes int64) (*repb.Digest, []byte) {
+	generatorMu.Lock()
```

Comment:
can you put this lock inside the Generator?

Reply:
Done.